modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1359,7 +1359,47 @@ void GDScriptAnalyzer::reduce_assignment(GDScriptParser::AssignmentNode *p_assig
 		push_error("Cannot assign a new value to a constant.", p_assignment->assignee);
 	}
 
-	if (!is_type_compatible(p_assignment->assignee->get_datatype(), p_assignment->assigned_value->get_datatype(), true)) {
+	Variant::Operator vop = Variant::Operator::OP_EQUAL;
+	switch (p_assignment->operation) {
+		case GDScriptParser::AssignmentNode::OP_NONE:
+			vop = Variant::Operator::OP_EQUAL;
+			break;
+		case GDScriptParser::AssignmentNode::OP_ADDITION:
+			vop = Variant::Operator::OP_ADD;
+			break;
+		case GDScriptParser::AssignmentNode::OP_SUBTRACTION:
+			vop = Variant::Operator::OP_SUBTRACT;
+			break;
+		case GDScriptParser::AssignmentNode::OP_MULTIPLICATION:
+			vop = Variant::Operator::OP_MULTIPLY;
+			break;
+		case GDScriptParser::AssignmentNode::OP_DIVISION:
+			vop = Variant::Operator::OP_DIVIDE;
+			break;
+		case GDScriptParser::AssignmentNode::OP_MODULO:
+			vop = Variant::Operator::OP_MODULE;
+			break;
+		case GDScriptParser::AssignmentNode::OP_BIT_SHIFT_LEFT:
+			vop = Variant::Operator::OP_SHIFT_LEFT;
+			break;
+		case GDScriptParser::AssignmentNode::OP_BIT_SHIFT_RIGHT:
+			vop = Variant::Operator::OP_SHIFT_RIGHT;
+			break;
+		case GDScriptParser::AssignmentNode::OP_BIT_AND:
+			vop = Variant::Operator::OP_BIT_AND;
+			break;
+		case GDScriptParser::AssignmentNode::OP_BIT_OR:
+			vop = Variant::Operator::OP_BIT_OR;
+			break;
+		case GDScriptParser::AssignmentNode::OP_BIT_XOR:
+			vop = Variant::Operator::OP_BIT_XOR;
+			break;
+	}
+
+	bool compatible = false;
+	get_operation_type(vop, p_assignment->assignee->get_datatype(), p_assignment->assigned_value->get_datatype(), compatible);
```

Comment:
`get_operation_type` won't check the assignment compatibility, only the operation. So if this is false the error should be that the operands have invalid type for the operation. Then we need to store the return value of `get_operation_type to test the assignment compatibility. Otherwise something like this would be valid:

```gdscript
var i := 3
i *= Vector2(1, 1)
```

Reply:
@vnen fixed!

Reply:
Fixed with 2 different errors cases

```gdscript
int_val *= Vector2()
# Cannot assign a value of type "Vector2" to a target of type "int"
```

```gdscript
int_val *= String()
# Invalid operands "int" and "String" for assignment operator.
```